### ISSUE TITLE
PiSSA, OLoRA, LoftQ and VeRA never reach a trainer: use_vera r=256 trains 524,288 params where VeRA would train 2,048

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#794]
## The defect

Four documented PEFT variants are validated by the schema and then dropped. Every trainer builds a plain, randomly initialised LoRA:

| Field (`config/schema.py`) | What the docs promise | What trains |

|---|---|---|

| `lora.init_strategy: pissa` (`:129`) | SVD init of A/B from the base weight | random-init LoRA |

| `lora.init_strategy: olora` / `lora.use_olora: true` (`:103`) | QR (orthonormal) init | random-init LoRA |

| `lora.init_strategy: loftq` + `loftq_iter` / `loftq_bits` (`:141-153`) | quantisation-aware init | random-init LoRA |

| `lora.use_vera: true` (`:95`) | shared frozen random matrices, tiny trainable vectors | a full LoRA at the same `r` |

## Evidence

The two helpers that know about these fields have **no caller** in `src/`:

```

$ grep -rln "peft_builder\|loftq_init" src/soup_cli --include=*.py

(no output)

```

- `utils/peft_builder.py:14` `build_peft_config` is the only code that emits `init_lora_weights` or selects `VeraConfig`. Even it never forwards `loftq` (`:61` handles only `pissa` / `olora`).

- `utils/loftq_init.py:52` `build_loftq_config` builds a `LoftQConfig` that nothing uses.

- `git log -S build_peft_config -- src/soup_cli/trainer src/soup_cli/commands` and `git log -S init_lora_weights -- src/soup_cli/trainer` both return nothing, so no trainer has ever called either.

What the trainers use instead:

- SFT and pretrain call `build_lora_config_kwargs` (`utils/peft_wiring.py:106-129`), which forwards `r`, `alpha`, `dropout`, `use_dora`, `use_rslora`, `rank_pattern` and `alpha_pattern`, but not `init_strategy`, `use_olora`, `use_vera` or `loftq_*`.

- Every other trainer constructs `LoraConfig(r=..., lora_alpha=..., use_dora=..., use_rslora=...)` inline, for example `dpo.py:317`, `grpo.py:621`, `kto.py:278`, `orpo.py:303`, `simpo.py:288`, `ipo.py:246`, `bco.py:300` and the SFT vision path at `sft.py:1655`. Of the 20 `LoraConfig(` constructions under `trainer/`, 2 go through the shared builder and 18 are inline.

The fields are not in `KNOWN_UNCONSUMED` in `tests/test_issue748_config_fields_reach_a_consumer.py`. The #748 guard counts the attribute reads inside `peft_builder.py`, which nothing imports, as consumption.

## Reproduction

The configs parse on every task tried (base `SmolLM2-135M`, `quantization: none`):

```

'init_strategy: pissa'                   sft   accepted (init_strategy=pissa, use_vera=False, r=64)

'init_strategy: pissa'                   dpo   accepted (init_strategy=pissa, use_vera=False, r=64)

'init_strategy: pissa'                   grpo  accepted (init_strategy=pissa, use_vera=False, r=64)

'use_olora: true'                        sft   accepted (init_strategy=olora, use_vera=False, r=64)

'use_olora: true'                        dpo   accepted (init_strategy=olora, use_vera=False, r=64)

'use_olora: true'                        grpo  accepted (init_strategy=olora, use_vera=False, r=64)

'init_strategy: loftq     loftq_bits: 4' sft   accepted (init_strategy=loftq, use_vera=False, r=64)

'init_strategy: loftq     loftq_bits: 4' dpo   accepted (init_strategy=loftq, use_vera=False, r=64)

'init_strategy: loftq     loftq_bits: 4' grpo  accepted (init_strategy=loftq, use_vera=False, r=64)

'r: 256     alpha: 1     use_vera: true' sft   accepted (init_strategy=random, use_vera=True, r=256)

'r: 256     alpha: 1     use_vera: true' dpo   accepted (init_strategy=random, use_vera=True, r=256)

'r: 256     alpha: 1     use_vera: true' grpo  accepted (init_strategy=random, use_vera=True, r=256)

```

Effect on a tiny random-init Llama (hidden 256, 2 layers, `target_modules: [q_proj, v_proj]`), going through the real SFT kwargs builder:

```

SFT kwargs keys: ['bias', 'lora_alpha', 'lora_dropout', 'r', 'target_modules', 'target_parameters', 'task_type', 'use_dora', 'use_rslora']

init_lora_weights in kwargs: False

soup path: base q_proj weight unchanged: True       # PiSSA did not run

real PiSSA: base q_proj weight unchanged: False     # PiSSA rewrites W into W_res

use_vera r=256 via soup path -> trainable params: 524288

real VeraConfig r=256        -> trainable params: 2048

```

## Why it matters

- PiSSA / OLoRA / LoftQ users get random init and read their loss curve as the method's result. LoftQ is specifically sold for QLoRA, where random init against a quantised base is the thing it exists to avoid.

- `use_vera` is the worst case. The docs (`docs/peft-and-efficiency.md:342-350`) recommend `r: 256` because VeRA needs a high rank. On the Soup path that becomes an `r=256` LoRA, **256x the trainable parameters** in the example above, with the matching optimizer state and VRAM. The adapter that was sold as "much smaller" comes out much larger.

- `docs/peft-and-efficiency.md` has worked examples for all four (`:200`, `:253`, `:349`, `:359`).

## Proposed fix

Route every trainer's LoRA construction through one builder. `build_lora_config_kwargs` in `peft_wiring.py` is already shared by SFT and pretrain, so the natural move is to make it return the class plus kwargs (absorbing `build_peft_config`), add `loftq_config` via `build_loftq_config`, and replace the 18 inline `LoraConfig(...)` sites with it. LoftQ also needs the base loaded un-quantised: peft's own docstring says "you should not pass a quantized model in this case, as LoftQ will quantize the model itself" (`peft/tuners/lora/config.py:463-465`), while Soup defaults to `quantization: 4bit`.

If that is too large for one change, reject non-default `init_strategy` / `use_vera` / `use_olora` at parse time for every task, with a message naming the field, until each trainer is wired. The schema already does exactly that for `stream_layers` (`schema.py:5283-5300`).

## Acceptance criteria

- [ ] A test builds a tiny model through a trainer's real LoRA path with `init_strategy: pissa` and asserts the base weight changed (PiSSA writes `W_res`). It must fail on current code.

- [ ] A test asserts that `use_vera: true` produces a `VeraModel` / VeRA layers, not `lora.Linear`.

- [ ] A test asserts `init_strategy: loftq` passes `init_lora_weights="loftq"` with a `LoftQConfig`, or is refused.

- [ ] A structural test fails if any `trainer/*.py` constructs `LoraConfig(` without going through the shared builder.

- [ ] Any task left unwired refuses these fields at parse time.

---

**Related, filed the same day:**

- #803 — lora.rank_pattern / alpha_pattern reach 2 of 20 LoRA construction sites: DPO and KTO train q_proj at r=16 w...
- #807 — The #748 consumer guard counts a read inside a function nothing calls: training.warmup_auto passes it and d...

**Comment by Amix29** [comment]
I will take this after #796. I will build on the shared LoRA wiring introduced for #803 so PiSSA, OLoRA, LoftQ and VeRA either reach every applicable trainer through one constructor or are refused explicitly at config load. Tests will exercise real PEFT config/model types on a tiny CPU model, pin LoftQ configuration and reject any remaining inline trainer LoraConfig construction; no CUDA path is needed for the correctness boundary.

**Comment by MakazhanAlpamys** [comment]
Closed by #1037 (`094de2bd`) from @Amix29 — thank you.

`lora.init_strategy` (pissa / olora / loftq) and `lora.use_vera` now reach the adapter on every trainer. Before, they were validated and then dropped, and `use_vera` quietly built a plain LoRA. Invalid combinations are refused at load:
- LoftQ or PiSSA on a pre-quantized base;
- VeRA on `backend: mlx`;
- any variant on `moe_lora_routing`.

Each strategy has a behavioural test through the real SFT setup that asserts its effect, not the config field.

Still open as a separate concern: pissa, olora and loftq rewrite the base weight, so the resulting adapter is not equivalent when served on the original base.

Not in a release yet; ships with the next minor.

Try: `init_strategy: pissa` under `training.lora` in a `task: sft` config with `quantization: none`.
